evap/staff/tests/test_views.py

```diff
@@ -1571,16 +1588,41 @@ class TestCourseCommentsUpdatePublishView(WebTest):
     @classmethod
     def setUpTestData(cls):
         mommy.make(UserProfile, username="staff.user", groups=[Group.objects.get(name="Staff")])
-        cls.course = mommy.make(Course)
+        cls.student1 = mommy.make(UserProfile)
+        cls.student2 = mommy.make(UserProfile)
+        cls.course = mommy.make(Course, participants=[cls.student1, cls.student2], voters=[cls.student1])
 
-    def helper(self, old_state, expected_new_state, action):
+    def helper(self, old_state, expected_new_state, action, expect_errors=False):
         textanswer = mommy.make(TextAnswer, state=old_state)
-        response = self.app.post(self.url, params={"id": textanswer.id, "action": action, "course_id": self.course.pk}, user="staff.user")
-        self.assertEqual(response.status_code, 200)
-        textanswer.refresh_from_db()
-        self.assertEqual(textanswer.state, expected_new_state)
+        response = self.app.post(self.url, params={"id": textanswer.id, "action": action, "course_id": self.course.pk}, user="staff.user", expect_errors=expect_errors)
+        if expect_errors:
+            self.assertEqual(response.status_code, 403)
+        else:
+            self.assertEqual(response.status_code, 200)
+            textanswer.refresh_from_db()
+            self.assertEqual(textanswer.state, expected_new_state)
+
+    def test_review_actions_without_publish_confirmation(self):
+        # in a course with only one voter reviewing should fail
+        self.helper(TextAnswer.NOT_REVIEWED, TextAnswer.PUBLISHED, "publish", expect_errors=True)
+
+        self.course.voters.add(self.student2)
+        self.course.save()
+
+        # now reviewing should work
+        self.helper(TextAnswer.NOT_REVIEWED, TextAnswer.PUBLISHED, "publish")
+        self.helper(TextAnswer.NOT_REVIEWED, TextAnswer.HIDDEN, "hide")
+        self.helper(TextAnswer.NOT_REVIEWED, TextAnswer.PRIVATE, "make_private")
+        self.helper(TextAnswer.PUBLISHED, TextAnswer.NOT_REVIEWED, "unreview")
+
+    def test_review_actions_with_publish_confirmation(self):
+        # in a course with only one voter reviewing should fail
+        self.helper(TextAnswer.NOT_REVIEWED, TextAnswer.PUBLISHED, "publish", expect_errors=True)
+
+        self.course.can_publish_text_results = True
```

Comment:
correct. meaning that both tests should use a helper class? meaning that we don't need the second test because there is another one testing that adding a second voter sets this to true?

Reply:
i don't quite understand since to me it looks like you added the duplicate test. yes, please remove it.